File: src/main/java/com/buzzvil/rxjava/sample/network/NetworkTest.java
package com.buzzvil.rxjava.sample.network;

import io.reactivex.Observable;
import io.reactivex.schedulers.Schedulers;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;
import retrofit2.http.GET;

import java.util.List;

public class NetworkTest {

    public static final String BASE_HOST = "https://jsonplaceholder.typicode.com/";

    public static void main(String[] args) {
        HttpService httpService = HttpService.retrofit.create(HttpService.class);
        Observable<List<Contents>> contentsObservable = httpService.requestPosts();
        contentsObservable
                .flatMapIterable(x -> x)
                .filter(contents -> contents.getId() == 99)
                .map(contents -> "Filtered Target : " + contents.getTitle())
                .observeOn(Schedulers.newThread())
                .subscribe(System.out::println);
    }

    public interface HttpService {
        @GET("posts")
        Observable<List<Contents>> requestPosts();

        Retrofit retrofit = new Retrofit.Builder()
                .baseUrl(BASE_HOST)
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .addConverterFactory(GsonConverterFactory.create())
                .build();
    }

    @Data
    @AllArgsConstructor
    @NoArgsConstructor
    public static class Contents {
        private int userId;
        private int id;
        private String title;
        private String body;
    }
}

File: src/main/kotlin/com/buzzvil/rxjava/sample/network/NetworkTest.kt
package com.buzzvil.rxjava.sample.network

import io.reactivex.Observable
import io.reactivex.schedulers.Schedulers
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory
import retrofit2.http.GET

fun main(args: Array<String>) {
    val httpService = retrofit.create(HttpService::class.java)
    httpService.requestPosts()
            .flatMapIterable { it }
            .filter { it.id == 99 }
            .map { "Filtered Target : ${it.title}" }
            .observeOn(Schedulers.newThread())
            .subscribe { println(it) }
}

interface HttpService {
    @GET("posts")
    fun requestPosts(): Observable<List<Contents>>
}

const val BASE_HOST = "https://jsonplaceholder.typicode.com/"
val retrofit: Retrofit = Retrofit.Builder()
        .baseUrl(BASE_HOST)
        .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
        .addConverterFactory(GsonConverterFactory.create())
        .build()

data class Contents(val userId: Int, val id: Int, val title: String, val body: String)
